### PR TITLE
never retry pod/build sync delete failures

### PR DESCRIPTION
**Pull request by bparees**

### CODE REVIEW

**Comment by bparees** [comment]
@liggitt @smarterclayton 
This might resolve https://github.com/openshift/origin/issues/2864, i'm guessing the state we got into is that we had deleted the build and we were trying to delete the associated pod but failed to fetch it and this logic caught an error instead of nil:
https://github.com/openshift/origin/blob/master/pkg/build/controller/controller.go#L279

which sent us into a retry loop to retry handling the build deletion basically forever.  What bothers me is i can't see how @liggitt got the particular error message he reported.  everything i'm seeing looks like it would come out as a different message.  In fact grepping the code i can't find anything that would throw a message that looked like that.

There's really no need to retry handling a build deletion, we can just let the sync loop deal with it.  in this scenario the syncloop will no-op as it will neither find a pod with an unmatched build, nor a build with an unmatched pod, and we'll be done.

With this new logic we've got a belt+suspenders:
1) we won't retry errors handling deletion events
2) we shouldn't throw an error if the pod doesn't exist for the build that's being deleted, we'll just silently accept that nothing needs to be done.

**Comment by bparees** [comment]
[test]

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func (c *RetryController) handleOne(resource interface{}) {
 		c.Retry(resource, err)
 		return
 	}
-	c.RetryManager.Forget(resource)
+	c.Forget(resource)
```
Comment (liggitt) [inline pkg/controller/controller.go]: was this change intentional, and what does it do? the comment needs updating if so

Comment (liggitt) [inline pkg/controller/controller.go]: oh, embedded, nm

File: pkg/build/controller/controller.go [hunk 2]
```diff
@@ -279,7 +279,7 @@ func (bc *BuildDeleteController) HandleBuildDeletion(build *buildapi.Build) erro
 	glog.V(4).Infof("Handling deletion of build %s", build.Name)
 	podName := buildutil.GetBuildPodName(build)
 	pod, err := bc.PodManager.GetPod(build.Namespace, podName)
-	if err != nil {
```
Comment (liggitt) [inline pkg/build/controller/controller.go]: also tolerate `IsNotFound` errors below when doing the actual deletion

File: pkg/build/controller/factory/factory.go [hunk 3]
```diff
@@ -112,7 +112,7 @@ func (factory *BuildControllerFactory) CreateDeleteController() controller.Runna
 		RetryManager: controller.NewQueueRetryManager(
 			queue,
 			cache.MetaNamespaceKeyFunc,
-			limitedLogAndRetry(factory.BuildUpdater, 30*time.Minute),
+			controller.RetryNever,
```
Comment (liggitt) [inline pkg/build/controller/factory/factory.go]: are you sure we don't want to retry on raised errors, now that NotFound errors don't propagate? what will clean up the build pod if not this?

**Comment by liggitt** [comment]
> What bothers me is i can't see how @liggitt got the particular error message he reported

It's a standard not found error:
```
func NewNotFound(kind, name string) error {
	return &StatusError{api.Status{
		Status: api.StatusFailure,
		Code:   http.StatusNotFound,
		Reason: api.StatusReasonNotFound,
		Details: &api.StatusDetails{
			Kind: kind,
			ID:   name,
		},
		Message: fmt.Sprintf("%s %q not found", kind, name),
	}}
}
```

File: pkg/build/controller/controller.go (hunk 2 above)
Comment (bparees) [inline pkg/build/controller/controller.go]: good call. adding.

File: pkg/build/controller/factory/factory.go (hunk 3 above)
Comment (bparees) [inline pkg/build/controller/factory/factory.go]: the sync loop will ultimately retry this whole thing if the state of "build pod with missing build" continues:
https://github.com/openshift/origin/blob/master/pkg/build/controller/factory/factory.go#L353

**Comment by bparees** [comment]
ah, thanks.  I was looking for something that actually started with "pods".  now i feel much better that this addresses what you hit though.

**Comment by liggitt** [comment]
I'd be happy with just treaing NotFound errors as a successful delete... I'm iffy on throwing in the towel on any other error and waiting for a sync loop. Is there any downside to the build pod possibly hanging around until the loop runs?

**Comment by bparees** [comment]
no downside other than some wasted resources/quota since the (if it was running) will continue to run longer.

**Comment by bparees** [comment]
(at least that i can think of at 2am)

**Comment by liggitt** [comment]
NotFound bits LGTM, maybe wait for the cold light of morning before merging the retry changes

**Comment by bparees** [comment]
@liggitt we can let @smarterclayton be the arbiter of whether retry should be excised.  but personally i kinda like the simpler approach here of no retrying anything and just letting the sync loop deal with it.

**Comment by liggitt** [comment]
works for me, I'm more unsure than opposed. though if we do want to remove retry logic, constructing a RetryManager with a RetryNever policy cracks me up

**Comment by liggitt** [comment]
you could even split the PR in twain, merge the NotFound bits to resolve the issue, and get more internet points

**Comment by bparees** [comment]
@smarterclayton you ok with this one?

File: pkg/build/controller/factory/factory.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/build/controller/factory/factory.go]: I don't think we should have failed the build anyway if the pod can't be cleaned up (by definition) so this is probably ok.

**Comment by bparees** [comment]
@smarterclayton @danmcp can i get a merge?

**Comment by smarterclayton** [comment]
@liggitt is that a LGTM?

**Comment by liggitt** [comment]
yes, LGTM

**Comment by smarterclayton** [comment]
Approved [merge]
